proposals/simd-0182-conditional-cu-metering.md

```diff
@@ -0,0 +1,114 @@
+of CUs consumed up to the point of failure requires additional effort. For
+instance, the Agave client implements a mechanism that tracks the Instruction
+Pointer (IP) or Program Counter (PC) to backtrack and estimate the CUs consumed
+when an exception occurs. More details on this mechanism can be found
+[here](https://github.com/solana-labs/rbpf/blob/57139e9e1fca4f01155f7d99bc55cdcc25b0bc04/src/jit.rs#L267).
+
+While this approach is effective, it introduces additional work and complexity.
+These mechanisms are often implementation-specific, and requiring all clients to
+track the exact number of executed sBPF instructions for consensus is costly and
+unnecessary. Such precision is not essential for protocol-level consensus,
+especially since these cases are rare.
+
+Instead of mandating implementation-specific work to handle exceptions, we
+propose the following clarification in the protocol:
+
+- For successful execution of a Basic Block (i.e., the block exits at the last
+ sBPF instruction), the deterministic CU cost of the block will be charged to
+the transaction’s CU meter. This ensures that CU consumption for successful
+transactions is accurately accounted for.
+- In the event of irregular failure, where execution aborts from the middle of
+basic block, the requested CUs for the transaction will be charged to the CU
+meter. This allows for a simple and efficient fallback mechanism that avoids the
+need for tracking the exact number of executed instructions up to the point of
+failure.
+
+By adopting this approach, the protocol avoids the overhead of requiring precise
+instruction-level CU tracking for transactions that fail. Instead, the requested
+CU limit of the transaction will be used, simplifying the handling of
+irregularly failed transactions while still maintaining consensus.
+
+## Alternatives Considered
+
+None
+
+## New Terminology
+
+- [Basic Block](https://en.wikipedia.org/wiki/Basic_block): In the context of
+  JIT execution and BPF processing, a Basic Block is a sequence of BPF
+instructions that forms a single, linear flow of control with no loops or
+conditional branches except for the entry and exit points. It represents a
+segment of code where execution starts at the first instruction and proceeds
+sequentially through to the last instruction without deviation. The Basic Block
+is characterized by its predictable execution path, allowing for efficient
+budget checks and optimizations, as its Compute Unit (CU) cost can be determined
+before execution and verified at the end of the block.
+
+- Irregular transaction failure: A rare case that a Transaction execution aborts
+in the middle of executing basic block, results in consuming all requested CUs.
+
+## Detailed Design
+
+If VM execution returns any error except `SyscallError`, transaction's CU meter
+should be fully depleted, in another words, all requested CUs are consumed;
+otherwise consumes the actual executed CUs.
```

Comment:
Just out of curiosity, why make syscall errors exempt?

Reply:
My understanding is syscall not happen in VM realm. @Lichtso @ptaffet-jump please correct me if I'm wrong.

Reply:
Syscalls should be included in this. Any error returned by the vm should cause the transaction meter to be fully depleted.

Reply:
Syscall errors are never thrown in the middle of basic blocks, so we can easily know exactly how many CUs were consumed when the error was thrown.

Reply:
In the usual definition of basic blocks, call instructions can occur anywhere in a basic block (apart from the last instruction). So I would say this is unclear.

Also, it does not explain the need for the exemption. I would expect all CUs to be all consumed if a program failed, including if that failure was in a syscall.

Reply:
> I would expect all CUs to be all consumed if a program failed, including if that failure was in a syscall.

Why would you expect that? If a program makes a CPI and that invoked program returns an error, we immediately stop executing the transaction and therefore have more time to execute other transactions in a block. If we consume all remaining CUs for that failure, we are wasting that capacity for execution.

Reply:
@seanyoung: This is mostly about how easy / hard it is to pinpoint the precise fault location. During syscalls the VM has stopped anyway and the return address is known on the stack. Otherwise it can be any location since the last metering (which happens at controlflow instructions). Also syscall errors are far more common and expected to happen regularly, whereas other faults such as access violation only happen in broken programs.

Reply:
@Lichtso that makes sense :+1: